lib/WTSI/NPG/HTS/LocationWriter.pm

```diff
@@ -0,0 +1,230 @@
+
+has 'pipeline_name' =>(
+  isa           => 'Str',
+  is            => 'ro',
+  required      => 1,
+  default       => $NPG_PROD,
+  documentation => 'The name of the pipeline used to produce the data');
+
+has 'platform_name' => (
+  isa           => 'Str',
+  is            => 'ro',
+  required      => 1,
+  documentation => 'Name of the platform used to produce raw data');
+
+sub _build_locations {
+  my ($self) = @_;
+  my $locations = {};
+  if (-e $self->path) {
+    open my $fh , '<:encoding(UTF-8)', $self->path or
+      $self->logcroak(q[could not open ml warehouse json file] .
+        $self->path);
+    my $file_contents = <$fh>;
+    close $fh or $self->logcroak(q[could not close ml warehouse ] .
+        qq[json file $self->path]);
+
+    try {
+      my $decoded = $self->decode($file_contents);
+      foreach my $product (@{$decoded->{products}}){
+        my ($key, $paths) = $self->_build_location_pair(
+          $product->{irods_root_collection},
+          $product->{irods_data_relative_path},
+          $product->{id_product},
+          $product->{irods_secondary_data_relative_path}
+          );
+        $locations->{$key} = $paths
+
+      }
+      $self->debug('Read previous locations from file: ', $self->path);
+
+    } catch{
+      $self->logcroak('Failed to parse locations from JSON file: ', $self->path);
+    }
+  }else{
+    $self->debug("No file at $self->{path}, using empty location arrayref");
+    $locations = {};
+  }
+  return $locations;
+
+}
+
+=head2 add_location
+
+  Named args : coll                 Collection. Str.
+               path                 Relative Path. Str.
+               pid                  Product Id. Str.
+               secondary_path       Path to secondary data object for this product
+                                    Str. Optional.
+
+  Example    : $self->add_location(coll     => $collection,
+                                   path     => $path,
+                                   pid      => $pid);
+  Description: Add a key value pair to the location hash, replacing any prior
+               value
+  Returntype : Void
+
+=cut
+sub add_location{
+  my $positional = 1;
+  my @named      = qw[pid coll path secondary_path];
+  my $params     = function_params($positional, @named);
+
+  my ($self) = $params->parse(@_);
+
+  #initialise location hash if not already done
+  $self->locations;
+
+  my $secondary_path = q[];
+  if ($params->secondary_path){
+    $secondary_path = $params->secondary_path;
+  }
+  my ($key, $paths) = $self->_build_location_pair(
+    $params->coll,
+    $params->path,
+    $params->pid,
+    $secondary_path);
+  $self->{locations}->{$key} = $paths;
+
+  return;
+}
+
+sub _build_location_pair{
+
+  my ($self, $coll, $path, $pid, $secondary_path) = @_;
+
+  if ($coll !~ m{.*/$}xms){
+    $coll .= q[/]
+  }
+
+  my $key = $coll . $DELIMITER . $pid;
+  my $paths = $path;
+  if ($secondary_path){
+    $paths .= $DELIMITER . $secondary_path;
```

Comment:
Could you get clashes if `$DELIMITER` exists in the paths?

Reply:
Yes, but that shouldn't happen with a comma, unless I'm missing something

Reply:
I think null was causing some test failures, but I'll rerun tests with it, in case I was mistaken

Reply:
It appears that I was wrong about what caused those failures